### PR TITLE
Remove all build warnings

### PR DESCRIPTION
**Pull request by calgray**
Here I've cleaned up the remaining build warnings by adding "-Werror" and defined the set of cuda warnings to ignore for all projects in the solution.

### CODE REVIEW

**Comment by calgray** [comment]
There is further type checking that can be made by removing c-style casts with `-Wold-style-cast` however I'd need to find a way to enable it only for gcc as cuComplex.h included in CUDA source files is c style.

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,19 +3,20 @@ cmake_minimum_required(VERSION 3.8)
...
 # Project configuration, specifying version, languages,
 # and the C++ standard to use for the whole project
-project(${SOLUTION_NAME} LANGUAGES CXX VERSION ${SOLUTION_NAME_VERSION})
+project(${SOLUTION_NAME} LANGUAGES CXX VERSION ${SOLUTION_VERSION})
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 if (CMAKE_CXX_COMPILER_ID MATCHES ".*Clang" OR
     CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Werror")
```
Comment (rtobar) [inline CMakeLists.txt:19]: Could we have an option (defaulting to `ON`) to include the extra `-Werror` flag? It's all good and well with leap-accelerate, but casacore or other third-party dependencies can give you warnings, and you'll have no control over that.

File: cmake/CUDAConfig.cmake [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+# CUDA Options
+if(CUDA_VERSION_STRING VERSION_GREATER 10)
+  # 2829 annotation on a defaulted function is ignored
+  # 3057 annotation is ignored on a function that is explicitly defaulted
+  # 2929 annotation is ignored on a function that is explicitly defaulted
+  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=3057,2929">)
+else()
+  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=2829">)
+endif()
```
Comment (rtobar) [inline cmake/CUDAConfig.cmake:10]: Having this in a separate file is great, but it doesn't get used. I'd wrap it into a function (called e.g., `apply_cuda_target_options`) that accepts `TARGET_NAME` as a parameter, include this file from the top-level `CMakeLists.txt` file, then call the function where you currently have this code duplicated.

File: src/icrar/leap-accelerate-cli/tests/CMakeLists.txt [hunk 3]
```diff
@@ -22,6 +22,16 @@ add_executable(${TEST_TARGET}
...
+# CUDA Options
+if(CUDA_VERSION_STRING VERSION_GREATER 10)
+  # 2829 annotation on a defaulted function is ignored
+  # 3057 annotation is ignored on a function that is explicitly defaulted
+  # 2929 annotation is ignored on a function that is explicitly defaulted
+  target_compile_options(${TEST_TARGET} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=3057,2929">)
+else()
+  target_compile_options(${TEST_TARGET} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=2829">)
+endif()
+
```
Comment (rtobar) [inline src/icrar/leap-accelerate-cli/tests/CMakeLists.txt:34]: Invoke the new function here, e.g. `apply_cuda_target_options(${TARGET_NAME})`.

File: src/icrar/leap-accelerate/CMakeLists.txt [hunk 4]
```diff
@@ -122,7 +122,8 @@ target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudaf
 if(CUDA_VERSION_STRING VERSION_GREATER 10)
   # 2829 annotation on a defaulted function is ignored
   # 3057 annotation is ignored on a function that is explicitly defaulted
-  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=3057">)
+  # 2929 annotation is ignored on a function that is explicitly defaulted
+  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=3057,2929">)
 else()
   target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=2829">)
 endif()
```
Comment (rtobar) [inline src/icrar/leap-accelerate/CMakeLists.txt:129]: Invoke the new function here, e.g. `apply_cuda_target_options(${TARGET_NAME})`.

File: src/icrar/leap-accelerate/tests/CMakeLists.txt [hunk 5]
```diff
@@ -31,28 +31,39 @@ if(CUDA_VERSION_MAJOR GREATER 9)
...
+target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe --display_error_number>)
+
+if(CUDA_VERSION_STRING VERSION_GREATER 10)
+  # 2829 annotation on a defaulted function is ignored
+  # 3057 annotation is ignored on a function that is explicitly defaulted
+  # 2929 annotation is ignored on a function that is explicitly defaulted
+  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=3057,2929">)
+else()
+  target_compile_options(${TARGET_NAME} PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcudafe="--diag_suppress=2829">)
+endif()
```
Comment (rtobar) [inline src/icrar/leap-accelerate/tests/CMakeLists.txt:56]: Invoke the new function here, e.g. `apply_cuda_target_options(${TARGET_NAME})`.

**Review by rtobar — COMMENTED** [review]
I checked locally and indeed got all compiling without warnings. I left only two main comments around code reuse, and turning errors optional (but on by default), otherwise it's great to see these changes.

File: CMakeLists.txt (hunk 1 above)
Comment (calgray) [reply CMakeLists.txt:19]: Tying this to an option sounds like a good idea. Ideally I just want to turn this on for release candidates.

File: src/icrar/leap-accelerate/core/logging.cc [hunk 6]
```diff
@@ -51,15 +51,15 @@ namespace log
...
         (
-            boost::log::trivial::severity >= boost::log::trivial::warning
+            boost::log::trivial::severity >= boost::log::trivial::info
         );
-        #else // Debug
+        #else // RELEASE
         boost::log::core::get()->set_filter
         (
-            boost::log::trivial::severity >= boost::log::trivial::info
+            boost::log::trivial::severity >= boost::log::trivial::trace
```
Comment (rtobar) [inline src/icrar/leap-accelerate/core/logging.cc:62]: I think you want `trace` in `DEBUG` and `info` on `RELEASE`?

Comment (calgray) [reply src/icrar/leap-accelerate/core/logging.cc:62]: true

**Review by rtobar — APPROVED** [review]

**Pull request merged**